moshi/src/main/java/com/squareup/moshi/Types.java

```diff
@@ -284,11 +284,21 @@ public static Type collectionElementType(Type context, Class<?> contextRawType)
     Type mapType = getSupertype(context, contextRawType, Map.class);
     if (mapType instanceof ParameterizedType) {
       ParameterizedType mapParameterizedType = (ParameterizedType) mapType;
-      return mapParameterizedType.getActualTypeArguments();
+      return getCanonicalizedMapTypeArguments(mapParameterizedType);
     }
     return new Type[] { Object.class, Object.class };
   }
 
+  private static Type[] getCanonicalizedMapTypeArguments(ParameterizedType parameterizedType) {
+    Type[] types = parameterizedType.getActualTypeArguments();
+    for (int i = 0; i < types.length; i++) {
+      if (types[i] instanceof GenericArrayType) {
+        types[i] = getRawType(types[i]);
```

Comment:
PROBABLY what we want to do is change the core Moshi loop so that when you call `Moshi.adapter()`, that’s where we convert one `byte[]` to the other. That’s a lot easier to do generally, and it means things that compose like `List` and `Map` will Just Work.